### PR TITLE
Add USDT tracepoints for key GC activities

### PR DESCRIPTION
**Pull request by caizixian**
This work is described in Improving Garbage Collection Observability with Performance Tracing by Claire Huang, Stephen M. Blackburn, and Zixian Cai, to appear in MPLR'23.

These tracepoints are compiled into `nop`s with minimal overhead when not attached as measured in the paper. The tracing tools that make use of these tracepoints will be released separately in another PR.

### CODE REVIEW

**Comment by caizixian** [comment]
The clippy warnings on macOS seems to be because the `probe!` macro compiles to an empty block on macOS. I think this is fine.

**Comment by qinsoon** [comment]
> The clippy warnings on macOS seems to be because the `probe!` macro compiles to an empty block on macOS. I think this is fine.

Yeah. Just suppress the warning.

File: src/plan/global.rs [hunk 1]
```diff
@@ -710,12 +710,14 @@ impl<VM: VMBinding> BasePlan<VM> {
     pub fn set_gc_status(&self, s: GcStatus) {
         let mut gc_status = self.gc_status.lock().unwrap();
         if *gc_status == GcStatus::NotInGC {
+            probe!(mmtk, gc_start);
```
Comment (qinsoon) [inline src/plan/global.rs:713]: Why not put this to the controller loop which is more reliable than `GcStatus`? 
https://github.com/mmtk/mmtk-core/blob/5d22c2b3ca6c0f3393f9c8cbd8b47a28868a48cc/src/scheduler/controller.rs#L54

We set `GcStatus::GcProper` in these places, and we have executed some packets before that.
https://github.com/mmtk/mmtk-core/blob/5d22c2b3ca6c0f3393f9c8cbd8b47a28868a48cc/src/scheduler/gc_work.rs#L452
https://github.com/mmtk/mmtk-core/blob/5d22c2b3ca6c0f3393f9c8cbd8b47a28868a48cc/src/scheduler/gc_work.rs#L475

File: src/scheduler/worker.rs [hunk 2]
```diff
@@ -350,7 +350,10 @@ impl<VM: VMBinding> GCWorker<VM> {
         self.tls = tls;
         self.copy = crate::plan::create_gc_worker_context(tls, mmtk);
         loop {
+            probe!(mmtk, work_poll);
```
Comment (qinsoon) [inline src/scheduler/worker.rs:353]: I was curious when I used the code: why use `work_poll` as a signal for work end. You can just have a work_end tracepoint after `do_work_with_stat`.

File: src/util/alloc/allocator.rs [hunk 3]
```diff
@@ -318,6 +318,19 @@ pub trait Allocator<VM: VMBinding>: Downcast {
...
     fn alloc_slow_once(&mut self, size: usize, align: usize, offset: usize) -> Address;
 
+    /// A wrapper method for [`alloc_slow_one`](Allocator::alloc_slow_once) to insert USDT tracepoints.
+    ///
+    /// Arguments:
+    /// * `size`: the allocation size in bytes.
+    /// * `align`: the required alignment in bytes.
+    /// * `offset` the required offset in bytes.
+    fn alloc_slow_once_traced(&mut self, size: usize, align: usize, offset: usize) -> Address {
+        probe!(mmtk, alloc_slow_once_start);
```
Comment (qinsoon) [inline src/util/alloc/allocator.rs:328]: These trace points are pretty clear. You can use the same pattern for gc start/end, and work start/end.

File: src/plan/global.rs (hunk 1 above)
Comment (caizixian) [reply src/plan/global.rs]: Fixed

File: src/scheduler/worker.rs (hunk 2 above)
Comment (caizixian) [reply src/scheduler/worker.rs:361]: Explained in the comment

File: src/util/alloc/allocator.rs (hunk 3 above)
Comment (caizixian) [reply src/util/alloc/allocator.rs:328]: Done

File: src/util/alloc/allocator.rs [hunk 4]
```diff
@@ -318,6 +318,19 @@ pub trait Allocator<VM: VMBinding>: Downcast {
     /// * `offset` the required offset in bytes.
     fn alloc_slow_once(&mut self, size: usize, align: usize, offset: usize) -> Address;
 
+    /// A wrapper method for [`alloc_slow_one`](Allocator::alloc_slow_once) to insert USDT tracepoints.
```
Comment (k-sareen) [inline src/util/alloc/allocator.rs:321]:
```suggestion
    /// A wrapper method for [`alloc_slow_once`](Allocator::alloc_slow_once) to insert USDT tracepoints.
```

**Comment by k-sareen** [comment]
Did you have the performance overheads of the extra indirection via `_trace` functions? I presume it's none since the compiler would inline it

**Comment by caizixian** [comment]
> Did you have the performance overheads of the extra indirection via `_trace` functions? I presume it's none since the compiler would inline it

I believe so.

**Comment by wks** [comment]
Can you also include an example bpftrace script somewhere in the `mmtk-core` repo?  For example, in `mmtk-core/examples/bpftrace/gcvis.bt`?  It would be more helpful if we also include a script that can convert its result into a trace format that can be opened by perfetto or other well-known tools.

**Comment by caizixian** [comment]
> Can you also include an example bpftrace script somewhere in the `mmtk-core` repo? For example, in `mmtk-core/examples/bpftrace/gcvis.bt`? It would be more helpful if we also include a script that can convert its result into a trace format that can be opened by perfetto or other well-known tools.

I plan to place the tracing scripts under `tool/` following the convention of bpftrace and bcc. 
I want to this in a separate PR, because I want to clean up the scripts to not have hardcoded paths etc.

**Comment by wks** [comment]
Can we also include USDT trace points for the following functions in mmtk-core?

-   `harness_begin`
-   `harness_end`
-   `GCController::run`
-   `GCWorker::run`

We usually need to initialise some states in the bpftrace trace.  Since mmtk-core doesn't provide a official C API (the bindings may create `extern "C"` functions if they need to), we can't directly use `uprobe` or `uretprobe` for those points.  We can simply add USDT in those functions so that we can trace those points in a VM-agnostic way.

**Review by wks — CHANGES_REQUESTED** [review]
I think this PR looks good in general.  The only missing peaces that I think we must include in this PR are the USDT trace points in the beginning of GC threads and `harness_{begin,end}`, without which it would be incomplete.

I am OK with postponing the splitting of `process_edges` trace point to another PR, because the current method is known to work.

I have another general suggestion, but doesn't have to be included in this PR, either.  We may let `GCWork` implement a very general `fn size(&self) -> Option<usize>` method so that when we do eBPF tracing, we can add a size argument to every work packet if it makes sense.  I think it makes sense for quite many packet types, including `ProcessEdgesWork`, `ScanObjects`, `Process{,Region}ModBuf` as well as weak reference and finalizer processing.  VM bindings can also use that to describe the size of their work packets, too, if that makes sense.  Examples include the number of "potential pinning parents" a Ruby VM has, the number of elements in weak global tables, such as "string tables" in OpenJDK and `fstring_table` in Ruby, etc.

File: src/scheduler/gc_work.rs [hunk 5]
```diff
@@ -631,6 +635,7 @@ pub trait ProcessEdgesWork:
     }
 
     fn process_edges(&mut self) {
+        probe!(mmtk, process_edges, self.edges.len(), self.is_roots());
```
[the remark this answers is not in the document]
Comment (caizixian) [reply src/scheduler/gc_work.rs:638]: Yeah, let's split into other PRs. We can always refine the tracepoints we have when we have a compelling use case.

**Review by wks — APPROVED** [review]
LGTM

**Pull request merged**